### PR TITLE
Adding hostname to groups discovery information

### PR DESCRIPTION
**Pull request by nikhiljindal**
Based on https://github.com/kubernetes/kubernetes/pull/20358#issuecomment-179577357

Ref https://github.com/kubernetes/kubernetes/issues/20229

Adding a publicHostname field to groups discovery information.

To be used by dynamic client talking to a cluster with federated api servers.

cc @bgrant0607 @lavalamp @smarterclayton @kubernetes/sig-api-machinery

### CODE REVIEW

**Comment by lavalamp** [comment]
Can you also add the master change needed to populate this field? I think you want that in 1.2, also.

File: pkg/api/unversioned/types.go [hunk 1]
```diff
@@ -298,6 +298,8 @@ const (
 // +protobuf.options.(gogoproto.goproto_stringer)=false
 type APIVersions struct {
 	TypeMeta `json:",inline"`
+	// public hostname of the server that is serving this group.
+	PublicHostname string `json:"publicHostname"`
```
Comment (bgrant0607) [inline pkg/api/unversioned/types.go]: I'd like to be consistent with service terminology, which is "External" rather than "Public":

https://github.com/kubernetes/kubernetes/blob/master/pkg/api/v1/types.go#L1740

Comment (bgrant0607) [inline pkg/api/unversioned/types.go]: And are we also going to want the port?

Comment (bgrant0607) [inline pkg/api/unversioned/types.go]: And path prefix?

Comment (bgrant0607) [inline pkg/api/unversioned/types.go]: And would we want to allow an IP?

Comment (nikhiljindal) [inline pkg/api/unversioned/types.go]: Will change to External.

Yes, it can have IP and port as well.

**Comment by nikhiljindal** [comment]
Have updated the PR to populate the field as well.

This is how it looks for me on local cluster:

localhost:8080/apis:
apis

localhost:8080/api:
api

**Comment by smarterclayton** [comment]
Do we want to call it public?  What does "public" mean?  Do we use "public"
anywhere else in the API?  Do we want to formalize "public" as meaning
"visible outside the cluster"?

**Comment by nikhiljindal** [comment]
You are right @smarterclayton. @bgrant0607 also said the same thing. He suggested renaming public to external (which is what we use in ServiceSpec: https://github.com/kubernetes/kubernetes/blob/master/pkg/api/v1/types.go#L1740).

But on discussing further with @lavalamp and @bgrant0607, we decided to replace the external address field with a CIDRToAddressMap.
Clients can match their IP with the CIDR and figure out the address that they should use.
This assumes that the subnets have exclusive CIDRs. Clients can use the address with the longest CIDR that they match.
I will update the PR accordingly.

cc @thockin @bprashanth

**Comment by nikhiljindal** [comment]
Pushed a new commit as per the comment above.

I am using ServiceClusterIPRange to determine the internal CIDR (not sure if thats the right thing to use).
This is what I see now:

localhost:8080/api:
api_cidr

localhost:8080/apis:
apis_cidr

**Comment by smarterclayton** [comment]
Wow.... that's... interesting.   It uh... is certainly flexible.... :)

**Comment by smarterclayton** [comment]
Will this map to IPv6 cleanly?

**Comment by lavalamp** [comment]
It looks like you have the wrong port on the external address, I thought 6443 was the one we *listen on*, but we actually should be advertising 443? I forget how it's connected.

File: pkg/api/unversioned/types.go [hunk 2]
```diff
@@ -321,6 +323,15 @@ type APIGroup struct {
 	// preferredVersion is the version preferred by the API server, which
 	// probably is the storage version.
 	PreferredVersion GroupVersionForDiscovery `json:"preferredVersion,omitempty"`
+	// a map of CIDR to address of the server that is serving this group. Clients can use the appropriate address as per the CIDR that they match.
+	CIDRToAddressMap []CIDRToAddress `json:"cidrToAddressMap"`
```
Comment (lavalamp) [inline pkg/api/unversioned/types.go]: @thockin we need you to bless this type.

File: pkg/master/master.go [hunk 3]
```diff
@@ -526,6 +537,18 @@ func (m *Master) InstallThirdPartyResource(rsrc *extensions.ThirdPartyResource)
 	apiGroup := unversioned.APIGroup{
 		Name:     group,
 		Versions: []unversioned.GroupVersionForDiscovery{groupVersion},
+		CIDRToAddressMap: []unversioned.CIDRToAddress{
```
Comment (lavalamp) [inline pkg/master/master.go]: Suggest factoring out a common function instead of duplicating this?

File: pkg/genericapiserver/genericapiserver.go [hunk 4]
```diff
@@ -734,17 +755,8 @@ func (s *GenericAPIServer) newAPIGroupVersion(groupMeta apimachinery.GroupMeta,
...
-	hostAndPort := s.externalHost
+	hostAndPort := s.ExternalHost
 	protocol := "https://"
-
-	// TODO: this is kind of messed up, we should just pipe in the full URL from the outside, rather
-	// than guessing at it.
-	if len(s.externalHost) == 0 && s.ClusterIP != nil {
-		host := s.ClusterIP.String()
-		if s.PublicReadWritePort != 0 {
-			hostAndPort = net.JoinHostPort(host, strconv.Itoa(s.PublicReadWritePort))
```
Comment (lavalamp) [inline pkg/genericapiserver/genericapiserver.go]: I don't think you can just delete this code?

Comment (lavalamp) [inline pkg/genericapiserver/genericapiserver.go]: nvm, I see you actually moved it.

File: pkg/genericapiserver/genericapiserver.go [hunk 5]
```diff
@@ -276,7 +276,7 @@ type GenericAPIServer struct {
 	RequestContextMapper     api.RequestContextMapper
 
 	// External host is the name that should be used in external (public internet) URLs for this GenericAPIServer
-	externalHost string
+	ExternalHost string
```
Comment (lavalamp) [inline pkg/genericapiserver/genericapiserver.go]: Change name to ExternalAddress maybe?

**Comment by lavalamp** [comment]
I believe CIDRs are still things for IPv6, so it should. Also this would
let us provide IPv6 and IPv4 addresses, which could be handy, though that
wasn't what we were thinking.

**Comment by smarterclayton** [comment]
I'm ok with the map, I don't know how humans will understand it (it's awful
clever).  But I'd rather have this than something not well thought through.

File: pkg/master/master.go (hunk 3 above)
Comment (nikhiljindal) [inline pkg/master/master.go]: Done

File: pkg/genericapiserver/genericapiserver.go (hunk 5 above)
Comment (nikhiljindal) [inline pkg/genericapiserver/genericapiserver.go]: Done

**Comment by nikhiljindal** [comment]
Updated to use ServiceReadWritePort instead of ReadWritePort which switched the address to use 443 instead of 6443 port.
Also pushed a new version as per comments.

File: pkg/api/unversioned/types.go (hunk 2 above)
Comment (bgrant0607) [inline pkg/api/unversioned/types.go]: How about ServerAddressByClientCIDR?

**Comment by nikhiljindal** [comment]
Have also added a PathPrefix field as discussed.
@smarterclayton will hopefully like that :)

Here is the discovery output now:
```
$ curl http://localhost:8080/apis/
{
  "kind": "APIGroupList",
  "apiVersion": "v1",
  "groups": [
    {
      "name": "extensions",
      "versions": [
        {
          "groupVersion": "extensions/v1beta1",
          "version": "v1beta1"
        }
      ],
      "preferredVersion": {
        "groupVersion": "extensions/v1beta1",
        "version": "v1beta1"
      },
      "cidrToAddressMap": [
        {
          "cidr": "10.0.0.0/24",
          "address": "10.0.0.1:443"
        },
        {
          "cidr": "0.0.0.0/0",
          "address": "172.17.129.7:443"
        }
      ],
      "pathPrefix": "/apis"
    }
  ]
}
```

```
$ curl http://localhost:8080/api/
{
  "kind": "APIVersions",
  "apiVersion": "v1",
  "versions": [
    "v1"
  ],
  "cidrToAddressMap": [
    {
      "cidr": "10.0.0.0/24",
      "address": "10.0.0.1:443"
    },
    {
      "cidr": "0.0.0.0/0",
      "address": "172.17.129.7:443"
    }
  ],
  "pathPrefix": "/api"
}
```

**Comment by bgrant0607** [comment]
We at least need "external" (public Internet) and intra-cluster (kube-proxy) addresses. The idea behind the CIDRs is that they'd also facilitate non-external (e.g., VPN) inter-cluster addresses, accesses by machines/VMs on the same physical/virtual network but outside the cluster, etc.

**Comment by nikhiljindal** [comment]
@k8s-bot test this issue: #20646

**Comment by smarterclayton** [comment]
I'm ok with the CIDRs, it was a bit of a shock, but I think it's a good
future proof.

**Comment by smarterclayton** [comment]
It aligns well with a plan I had for ingress / routes to have the routers
identify which subnets they answer onto the route.

**Comment by nikhiljindal** [comment]
@k8s-bot test this issue: https://github.com/kubernetes/kubernetes/issues/20418

**Comment by bprashanth** [comment]
> It aligns well with a plan I had for ingress / routes to have the routers
identify which subnets they answer onto the route.

Yes I was thinking something along those lines

**Comment by lavalamp** [comment]
I don't think we need the prefix in the discovery information! The client
knows what URL it's asking for-- the prefix (and any proxy-imposed prefix)
has already been added!

**Comment by lavalamp** [comment]
The only place the prefix needs to appear is in the discoverizer's config, IMO. @smarterclayton to confirm.

...and either baked-in to clients, or at some given "what are the available prefixes" location. But not here.

**Comment by smarterclayton** [comment]
Right, summarizer output - no prefix, summarizer input config - has prefix

File: pkg/api/unversioned/types.go [hunk 6]
```diff
@@ -300,6 +300,10 @@ type APIVersions struct {
 	TypeMeta `json:",inline"`
 	// versions are the api versions that are available.
 	Versions []string `json:"versions"`
+	// a map of CIDR to address of the server that is serving this group. Clients can use the appropriate address as per the CIDR that they match.
```
Comment (thockin) [inline pkg/api/unversioned/types.go]: Is this CIDR for client IP or a destination IP?  The description is not clear.  Is it the client that does the mapping or do you map it server-side?

File: pkg/genericapiserver/genericapiserver.go [hunk 7]
```diff
@@ -698,6 +716,20 @@ func (s *GenericAPIServer) installAPIGroup(apiGroupInfo *APIGroupInfo) error {
 	return nil
 }
 
+func (s *GenericAPIServer) GetCIDRToAddressMap() []unversioned.CIDRToAddress {
+	return []unversioned.CIDRToAddress{
+		{
+			CIDR:    s.ServiceClusterIPRange.String(),
```
Comment (thockin) [inline pkg/genericapiserver/genericapiserver.go]: is this right?  The ServiceClusterIPRange is the range of addresses assigned to Services

File: pkg/api/unversioned/types.go [hunk 8]
```diff
@@ -329,6 +338,23 @@ type APIGroup struct {
...
+	ServerAddressByClientCIDRMap []ServerAddressByClientCIDR `json:"serverAddressByClientCIDRMap"`
+	// Optional path prefix that is inserted between the address and the group name to reach the group endpoint.
+	PathPrefix string `json:"pathPrefix,omitempty"`
+}
+
+// ServerAddressByClientCIDR helps the client to determine the server address that they should use, depending on the clientCIDR that they match.
+type ServerAddressByClientCIDR struct {
+	// The CIDR with which clients can match their IP to figure out the server address that they should use.
+	ClientCIDR string `json:"clientCIDR"`
+	// Address of this server, suitable for a client that matches the above CIDR. This can be a hostname or IP with port.
```
Comment (bgrant0607) [inline pkg/api/unversioned/types.go]: Specify exact syntax. Presumably hostname, IP, hostname:port, or IP:port?

Comment (smarterclayton) [inline pkg/api/unversioned/types.go]: Hostname or IP, with port.  I don't see how we can not support both.

Comment (nikhiljindal) [inline pkg/api/unversioned/types.go]: Fixed

File: pkg/api/unversioned/types.go (hunk 6 above)
Comment (nikhiljindal) [inline pkg/api/unversioned/types.go]: Updated the comment

File: pkg/genericapiserver/genericapiserver.go (hunk 7 above)
Comment (nikhiljindal) [inline pkg/genericapiserver/genericapiserver.go]: We are using it to approximate the client IPs that should be considered internal to the cluster.

**Comment by nikhiljindal** [comment]
Have updated the code as per discussion.
Have renamed the struct to ServerAddressByClientCIDR and updated the master code to look for the client IP and return internal CIDR only if the clientIP can match that.

Am looking for the client IP using X-Forwarded-For header first. If its empty, then I look at X-Real-Ip and if thats empty too then I use request.RemoteAddr.

PTAL. I can squash commits if its fine

File: pkg/api/unversioned/types.go [hunk 9]
```diff
@@ -308,6 +308,15 @@ type APIVersions struct {
...
 	Versions []string `json:"versions"`
+	// a map of client CIDR to server address that is serving this group.
+	// Clients can use the appropriate server address as per the CIDR that they match.
+	// In case of multiple matches, clients should use the longest matching CIDR.
+	// The server returns only those CIDRs that it thinks that the client can match.
+	// For ex: the master will return an internal IP CIDR only, if the client reaches the server using an internal IP.
+	// Server looks at X-Forwarded-For header or X-Real-Ip header or request.RemoteAddr (in that order) to get the client IP.
+	ServerAddressByClientCIDRMap []ServerAddressByClientCIDR `json:"serverAddressByClientCIDRMap"`
+	// Optional path prefix that is inserted between the address and the group name to reach the group endpoint.
+	PathPrefix string `json:"pathPrefix,omitempty"`
```
Comment (lavalamp) [inline pkg/api/unversioned/types.go]: @smarterclayton and I agreed elsewhere in this PR that this only needs to be in the summarizer's config, not here?

File: pkg/genericapiserver/genericapiserver.go [hunk 10]
```diff
@@ -224,9 +224,12 @@ type Config struct {
 	// The range of IPs to be assigned to services with type=ClusterIP or greater
 	ServiceClusterIPRange *net.IPNet
 
-	// The IP address for the GenericAPIServer service (must be inside ServiceClusterIPRange
+	// The IP address for the GenericAPIServer service (must be inside ServiceClusterIPRange)
 	ServiceReadWriteIP net.IP
 
+	// Port for the GenericAPIServer service.
```
Comment (lavalamp) [inline pkg/genericapiserver/genericapiserver.go]: Not generic apiserver-- the specific apiserver. E.g., `kubernetes`, `ubernetes`, `mypluginapiserver`

Comment (lavalamp) [inline pkg/genericapiserver/genericapiserver.go]: It's here because you want such a service for your specific apiserver.

File: pkg/master/master_test.go [hunk 11]
```diff
@@ -285,9 +285,12 @@ func TestDiscoveryAtAPIS(t *testing.T) {
...
 	}
-	assert.Equal(extensionsGroupName, groupList.Groups[0].Name)
-	assert.Equal(extensionsVersions, groupList.Groups[0].Versions)
-	assert.Equal(extensionsPreferredVersion, groupList.Groups[0].PreferredVersion)
+	assert.Equal(1, len(groupList.Groups))
+	groupListGroup := groupList.Groups[0]
+	assert.Equal(extensionsGroupName, groupListGroup.Name)
+	assert.Equal(extensionsVersions, groupListGroup.Versions)
+	assert.Equal(extensionsPreferredVersion, groupListGroup.PreferredVersion)
+	assert.Equal(master.APIGroupPrefix, groupListGroup.PathPrefix)
```
Comment (lavalamp) [inline pkg/master/master_test.go]: I had to fix this test in the batch move, it inappropriately assumes no other groups exist... I guess I'll just fix it again when I rebase.

**Comment by nikhiljindal** [comment]
Re: path prefix in discovery information, the reason for adding that was to let servers expose groups at other prefixes than `/apis` and still allow the summarizer to summarize all those groups (even the ones at different prefixes) at `/apis`. So for example: there are 2 servers, one exposing its groups at `/apis` and the other at `/hapis`, summarizer would list groups from both the servers at `/apis` and tell the clients that they need to go to `/hapis` on the second server.

@lavalamp suggested that the summarizer should only summarize groups at the same prefix. For ex: it should list groups from the first server at `/apis` and list the groups from second server at `/hapis`. If there is a third server in the cluster exposing groups at `/apis`, then the summarizer will present groups from both first and third server at `/apis`.
If we do that, then I think, we do need to add another discovery endpoint which lists all the possible prefixes in the cluster (`/apis`, `/hapis`, etc). Without that an admin cant just bring up a federated server in the cluster, configure the summarizer to include that server and expect that all clients will automatically discover the new server.

Will remove path prefix from this PR and discuss it separately.

**Comment by nikhiljindal** [comment]
Filed https://github.com/kubernetes/kubernetes/issues/20933 to discuss the path prefix issue.

**Comment by smarterclayton** [comment]
I don't like the "admin brings up server and APIs are magically
discovered" use case much - I expect admins to tightly control the
list of APIs exposed, *especially* if we do some sort of "front
authorization proxy".

If we have a config, the config should define the API groups.  We
don't need config *and* a second level of discovery, that just seems
like overkill.  An admin adding one config entry (/apis) vs two
(/apis/group1 and /apis/group2) is not a burden for the admin, and the
latter doesn't require us to maintain a discovery discovery code path,
or explain it to admins, or secure it, etc.

**Comment by nikhiljindal** [comment]
Sorry looks like I didnt explain myself well.
I totally agree with your point.

Cluster admin controls the summarizer config and he needs to update it for a new server or apigroup in the cluster to be exposed.

But my point was that even when the admin does that (he updates the summarizer config, so that it now includes the new groups), how does the client know that now there is a new path in the cluster (`<discoverySummarizerIP>/hapis`) that the client should talk to?
Note that this is only for the case when the new group is exposed at a different prefix then `/apis` (`/hapis` in the example above). 
It is possible that this is not a valid case that we want to support as per https://github.com/kubernetes/kubernetes/issues/20933#issuecomment-182104856

File: pkg/api/unversioned/types.go (hunk 9 above)
Comment (nikhiljindal) [inline pkg/api/unversioned/types.go]: Removed

File: pkg/genericapiserver/genericapiserver.go (hunk 10 above)
Comment (nikhiljindal) [inline pkg/genericapiserver/genericapiserver.go]: Changed to "Port for the apiserver service" :)

File: pkg/master/master_test.go (hunk 11 above)
Comment (nikhiljindal) [inline pkg/master/master_test.go]: Ack. Pull out the text fix in a separate PR?

**Comment by nikhiljindal** [comment]
Pushed a new version as per comments and squashed commits.
PTAL.

File: pkg/util/net/http.go [hunk 12]
```diff
@@ -120,3 +120,28 @@ func GetHTTPClient(req *http.Request) string {
...
+// Extracts and returns the clients IP from the given request.
+func GetClientIP(req *http.Request) (net.IP, error) {
+	hdr := req.Header
+	// First check the X-Forwarded-For header for requests via proxy.
+	hdrForwardedFor := hdr.Get("X-Forwarded-For")
+	if hdrForwardedFor != "" {
+		// X-Forwarded-For can be a csv of IPs in case of multiple proxies.
+		// Use the first one.
+		parts := strings.Split(hdrForwardedFor, ",")
+		ip := net.ParseIP(strings.TrimSpace(parts[0]))
```
Comment (lavalamp) [inline pkg/util/net/http.go]: You need to check that ip is not nil. Can you add a test of this?

File: pkg/util/net/http.go [hunk 13]
```diff
@@ -120,3 +120,28 @@ func GetHTTPClient(req *http.Request) string {
 	}
 	return "unknown"
 }
+
+// Extracts and returns the clients IP from the given request.
+func GetClientIP(req *http.Request) (net.IP, error) {
```
Comment (lavalamp) [inline pkg/util/net/http.go]: If you never return an error, maybe don't have it as a return parameter, just document that this is "best effort" therefore no error, but no promise of accuracy either.

File: pkg/util/net/http.go [hunk 14]
```diff
@@ -120,3 +120,28 @@ func GetHTTPClient(req *http.Request) string {
...
+	hdr := req.Header
+	// First check the X-Forwarded-For header for requests via proxy.
+	hdrForwardedFor := hdr.Get("X-Forwarded-For")
+	if hdrForwardedFor != "" {
+		// X-Forwarded-For can be a csv of IPs in case of multiple proxies.
+		// Use the first one.
```
Comment (lavalamp) [inline pkg/util/net/http.go]: @thockin do you see any problem with this approach?

File: pkg/api/unversioned/types.go [hunk 15]
```diff
@@ -329,6 +336,22 @@ type APIGroup struct {
...
+	// Server looks at X-Forwarded-For header or X-Real-Ip header or request.RemoteAddr (in that order) to get the client IP.
+	ServerAddressByClientCIDRMap []ServerAddressByClientCIDR `json:"serverAddressByClientCIDRMap"`
+}
+
+// ServerAddressByClientCIDR helps the client to determine the server address that they should use, depending on the clientCIDR that they match.
+type ServerAddressByClientCIDR struct {
+	// The CIDR with which clients can match their IP to figure out the server address that they should use.
+	ClientCIDR string `json:"clientCIDR"`
+	// Address of this server, suitable for a client that matches the above CIDR.
+	// This can be a hostname, hostname:port, IP or IP:port.
```
Comment (lavalamp) [inline pkg/api/unversioned/types.go]: Will we ever omit the port? I think we have to include it all the time?

File: pkg/api/unversioned/types.go [hunk 16]
```diff
@@ -308,6 +308,13 @@ type APIVersions struct {
 	TypeMeta `json:",inline"`
 	// versions are the api versions that are available.
 	Versions []string `json:"versions"`
+	// a map of client CIDR to server address that is serving this group.
+	// Clients can use the appropriate server address as per the CIDR that they match.
+	// In case of multiple matches, clients should use the longest matching CIDR.
+	// The server returns only those CIDRs that it thinks that the client can match.
+	// For ex: the master will return an internal IP CIDR only, if the client reaches the server using an internal IP.
```
Comment (lavalamp) [inline pkg/api/unversioned/types.go]: s/ex/example/

(or s/For ex:/E.g.:/)

File: pkg/api/unversioned/types.go [hunk 17]
```diff
@@ -308,6 +308,13 @@ type APIVersions struct {
 	TypeMeta `json:",inline"`
 	// versions are the api versions that are available.
 	Versions []string `json:"versions"`
+	// a map of client CIDR to server address that is serving this group.
```
Comment (lavalamp) [inline pkg/api/unversioned/types.go]: To help clients reach servers in the most network-efficient way possible, a map of client CIDR to a server address believed reachable and ideal from that CIDR.

File: pkg/api/unversioned/types.go [hunk 18]
```diff
@@ -308,6 +308,13 @@ type APIVersions struct {
...
+	// Clients can use the appropriate server address as per the CIDR that they match.
+	// In case of multiple matches, clients should use the longest matching CIDR.
+	// The server returns only those CIDRs that it thinks that the client can match.
+	// For ex: the master will return an internal IP CIDR only, if the client reaches the server using an internal IP.
+	// Server looks at X-Forwarded-For header or X-Real-Ip header or request.RemoteAddr (in that order) to get the client IP.
+	ServerAddressByClientCIDRMap []ServerAddressByClientCIDR `json:"serverAddressByClientCIDRMap"`
```
Comment (lavalamp) [inline pkg/api/unversioned/types.go]: @thockin or @bgrant0607 to check the API in this file, please, before we merge?

File: pkg/util/net/http.go (hunk 12 above)
Comment (nikhiljindal) [inline pkg/util/net/http.go]: Added test.
Also updated the func comment to explicitly say that it can return nil.
The current code in genericserver handles it fine. cidr.Matches() can handle nil.

File: pkg/util/net/http.go (hunk 13 above)
Comment (nikhiljindal) [inline pkg/util/net/http.go]: done.

File: pkg/api/unversioned/types.go (hunk 15 above)
Comment (nikhiljindal) [inline pkg/api/unversioned/types.go]: I guess we can omit it if its 80?

File: pkg/api/unversioned/types.go (hunk 16 above)
Comment (nikhiljindal) [inline pkg/api/unversioned/types.go]: done

File: pkg/api/unversioned/types.go (hunk 17 above)
Comment (nikhiljindal) [inline pkg/api/unversioned/types.go]: added

**Comment by nikhiljindal** [comment]
I have updated code and pushed a new version as per comments.
PTAL.

**Comment by bgrant0607** [comment]
The API looks fine. I might rename ServerAddressByClientCIDRMap to ServerAddressByClientCIDRs, but don't have a strong preference.

**Comment by smarterclayton** [comment]
I like the latter a bit more m - we don't say map much but By pretty
much always means map

File: pkg/api/unversioned/types.go (hunk 15 above)
Comment (lavalamp) [inline pkg/api/unversioned/types.go]: Won't it always be https? so 443 :) But yeah, I guess it's not a big deal.

File: pkg/util/net/http_test.go [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+				Header: map[string][]string{
+					"X-Forwarded-For": {ipString},
+				},
+			},
+			ExpectedIP: ip,
+		},
+		{
+			Request: http.Request{
+				Header: map[string][]string{
+					"X-Forwarded-For": {invalidIPString},
```
Comment (lavalamp) [inline pkg/util/net/http_test.go]: Add a case: `"X-Forwarded-For": {invalidIPString}, "X-Real-Ip": {ipString},`

I think this should return `ip`, not nil as I think it currently does?

File: pkg/util/net/http.go (hunk 12 above)
Comment (lavalamp) [inline pkg/util/net/http.go]: If it's nil, I think you should keep going on this function. That's what I was trying to say.

File: pkg/util/net/http.go [hunk 20]
```diff
@@ -120,3 +120,30 @@ func GetHTTPClient(req *http.Request) string {
...
+		parts := strings.Split(hdrForwardedFor, ",")
+		ip := net.ParseIP(strings.TrimSpace(parts[0]))
+		return ip
+	}
+
+	// Try the X-Real-Ip header if X-Forwarded-For is not set.
+	hdrRealIp := hdr.Get("X-Real-Ip")
+	if hdrRealIp != "" {
+		ip := net.ParseIP(hdrRealIp)
+		return ip
```
Comment (lavalamp) [inline pkg/util/net/http.go]: If ip is nil here, I think you should keep going also.

Comment (nikhiljindal) [inline pkg/util/net/http.go]: Done

File: pkg/util/net/http_test.go (hunk 19 above)
Comment (nikhiljindal) [inline pkg/util/net/http_test.go]: Done

**Comment by nikhiljindal** [comment]
Renamed ServerAddressByClientCIDRMap to ServerAddressByClientCIDRs and pushed a new version as per comments.
PTAL

**Comment by lavalamp** [comment]
LGTM

**Comment by nikhiljindal** [comment]
Rebased. Readding the LGTM tag

**Comment by nikhiljindal** [comment]
Added the release note label
